### PR TITLE
Add teacher-metrics to agent observations for display_model.

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
Adds teacher metrics into the model observation. This has the nice benefit that world logger and display model have access to this information and can log it.

**Testing steps**
New CI's covering the functionality.  Manual testing:

```
$ parlai dm -t convai2 -mf zoo:blender/blender_90M/model --inference topk --skip-generation false -v
[id]: convai2
[text]: my dad was always busy working at home depot
[eval_labels]: now that i am older home depot is my toy r us .
[label_candidates]: fun , i should take my kids out to do that . do you have a family ?|they were popular in my younger days lol|i sing and love rock especially while working . keeps me pumped .|i mean where i come from kitties do not play with birds .|i tend my rose garden and read books . i am deaf in one ear|... (5 of 20 shown)
   [id]: TransformerGenerator
   [beam_texts]: ("that ' s cool , what kind of things do they sell there ? i am looking for new shoes .", -13.5170259475708)
  ('my dad used to make money at home too , just not kosher food . . . lol', -14.683135032653809)
  ("i do not know what home depot is . but i ' ll have you know it is kosher !", -14.926774978637695)
  ("that is nice of him to do . did they have kosher food there ? that ' s what i ' d like to avoid .", -15.364789962768555)
  ('cool , i work at a grocery store , i love shopping . i only shop kosher though .', -15.92314338684082)
  ... (5 of 10 shown)
   [text]: that ' s cool , what kind of things do they sell there ? i am looking for new shoes .
   [metrics]:
    accuracy    bleu-4  exs    f1  loss   ppl  token_acc
           0 8.513e-08    1 .2069 3.236 25.44      .4286
   [token_losses]: now 6.837 | that 1.492 | i 2.398 | am 1.789 | older 2.599 | home 7.079 | depot 0.1772 | is 0.4172 | my 1.545 | toy 10.55 | r 9.772 | us 0.06661 | . 0.5078 | __end__ 0.07798 | __null__ 0.0 | __null__ 0.0
- - - - - - - END OF EPISODE - - - - - - - - - -
```

### CODE REVIEW

File: parlai/core/teachers.py [hunk 1]
```diff
@@ -448,10 +448,20 @@ def observe(self, observation):
...
         """
+        self.metrics.clear_recent()
         if hasattr(self, 'lastY') and self.lastY is not None:
             self.metrics.evaluate_response(observation, self.lastY)
             self.custom_evaluation(self.last_act, self.lastY, observation)
             self.lastY = None
+        recent_metrics = self.metrics.report_recent()
+        if recent_metrics:
+            # for display purposes (display_model), take all accumulated
+            # metrics back into the original observation. This is an abuse of
```
Comment (mojtaba-komeili) [inline parlai/core/teachers.py:459]: If it is an abuse, is there a better way of doing it later to add as a TODO?

File: tests/test_metrics.py [hunk 2]
```diff
@@ -179,6 +179,24 @@ def test_largebuffer(self):
...
+        assert m.report() == {'test': 3}
+        assert m.report_recent() == {'test': 2}
+        assert m2.report() == {'test': 3}
+        assert m2.report_recent() == {}
+        m2.add('test', SumMetric(3))
+        assert m2.report() == {'test': 6}
+        assert m.report() == {'test': 6}
+        assert m2.report_recent() == {'test': 3}
+        assert m.report_recent() == {'test': 2}
+
```
Comment (mojtaba-komeili) [inline tests/test_metrics.py:199]: I would add a `clear` and `clear_recent` for both `m` and `m2` at the end of this, 99.9% not needed, but doesn't heart.

File: parlai/core/teachers.py [hunk 3]
```diff
@@ -448,10 +448,20 @@ def observe(self, observation):
...
+            # Messages being pointers
+            if 'metrics' in observation:
+                # override agent-level metrics if present
+                observation.pop('metrics')
+            observation['metrics'] = recent_metrics
```
Comment (mojtaba-komeili) [inline parlai/core/teachers.py:464]: I do not understand why you pop the metric and then set it again. Could you explain this for me please?

**Review by mojtaba-komeili — APPROVED** [review]

File: parlai/core/teachers.py (hunk 1 above)
Comment (stephenroller) [reply parlai/core/teachers.py:459]: I'm not aware of one, otherwise I would have just done it.

File: tests/test_metrics.py (hunk 2 above)
Comment (stephenroller) [reply tests/test_metrics.py:199]: You mean with corresponding asserts to check state?

File: parlai/core/teachers.py (hunk 3 above)
Comment (stephenroller) [reply parlai/core/teachers.py:464]: https://github.com/facebookresearch/ParlAI/blob/840843185f292669814fc04272de41af74af4bb6/parlai/core/torch_agent.py#L2017-L2027

We send "local metrics" from TorchAgent (see [here](https://parl.ai/docs/tutorial_metrics.html#local-metrics)) to the teacher in the observation using this same key. This lets us distinguish metrics provenance when we are using Multitasking with mixed batches: every teacher observes the metrics of its example.

We then accumulate those local metrics inside the teacher, along with any metrics the teacher computes in this method: 

https://github.com/facebookresearch/ParlAI/blob/840843185f292669814fc04272de41af74af4bb6/parlai/core/metrics.py#L773-L822

Now that we want to include a dictionary of both teacher AND user agent metrics (for display and logging purposes), we need to replace that old metrics dict from TorchAgent. We could "update" it, but just overwriting it seems better imho

File: tests/test_metrics.py (hunk 2 above)
Comment (mojtaba-komeili) [reply tests/test_metrics.py:199]: Yes.

File: parlai/core/teachers.py (hunk 3 above)
Comment (mojtaba-komeili) [reply parlai/core/teachers.py:464]: Thanks for the explanation.

**Pull request merged**